### PR TITLE
ceph: add support for update() from lib-bucket-provisioner

### LINKED ISSUE #7146: Bucket quota can't be updated
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When creating `ObjectBucketClaim` user can specify `.spec.additionalConfig` values like `maxSize` and `maxObjects`. This value is enforced as a quota on the bucket. However if user later decides to update the spec and increase the quota, this change doesn't trigger reconciliation. New quota is not propagated to `ObjectBucket` resource or to the Ceph back-end bucket.

**Expected behavior:**
Quota should be updated.

**How to reproduce it (minimal and precise):**
1. Create`ObjectBucketClaim` with `maxSize` set to a value
2. `oc patch obc/RESOURCE_NAME -p '{"spec": {"additionalConfig": {"maxSize": "different value"}}
3. Compare `ObjectBucketClaim` and `ObjectBucket` and the actual quota. `maxSize` is changed only in `ObjectBucketClaim`

**File(s) to submit**:
n/a

**Environment**:
* OS (e.g. from /etc/os-release): RHEL
* Kernel (e.g. `uname -a`): 4.18
* Cloud provider or hardware configuration: Openshift 4.5
* Rook version (use `rook version` inside of a Rook Pod): `4.5-55.6dba87fd.release_4.5`
* Storage backend version (e.g. for ceph do `ceph -v`): `ceph version 14.2.8-111.el8cp (2e6029d57bc594eceba4751373da6505028c2650) nautilus (stable)`
* Kubernetes version (use `kubectl version`): client `v1.19.4`, server `v1.18.3`
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Openshift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): `HEALTH_OK`

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**
Recently lib-bucket-provisioner add support for update() API.
Include that on the obc implementation since it can be used to
update quota for OBC.

**Which issue is resolved by this Pull Request:**
Resolves #7146

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 1]
```diff
@@ -665,3 +665,69 @@ func (p *Provisioner) setAdminOpsAPIClient() error {
 
 	return nil
 }
+func (p Provisioner) updateAdditionalSettings(ob *bktv1alpha1.ObjectBucket) error {
+	var maxObjectsInt64 int64
+	var maxSizeInt64 int64
+	var err error
+	var quotaEnabled bool
+	maxObjects := MaxObjectQuota(ob.Spec.Endpoint.AdditionalConfigData)
+	maxSize := MaxSizeQuota(ob.Spec.Endpoint.AdditionalConfigData)
```
Comment (travisn) [inline pkg/operator/ceph/object/bucket/provisioner.go:674]: A lot of this method seems like duplicate code from `setAdditionalSettings()` above. Can this be refactored so creation and update call the same helper to process the additional settings?

Comment (thotz) [reply pkg/operator/ceph/object/bucket/provisioner.go:674]: @travisn : I agree that the variables used in the both functions are the same but checks used in both are different.
`setAdditionalSettings` will enable quota if `MaxSize` and `MaxObjects` are set, but `updateAdditionalSettings` checks condition for disabling quota and will enable quota only if existing value changes.

Comment (travisn) [reply pkg/operator/ceph/object/bucket/provisioner.go:674]: Ok I see the differences now. We can keep it separate for now.

**Review by travisn — COMMENTED** [review]
Please rebase on the latest master to see if the other CI fixes will help, or if there is still a CI issue to investigate. The object tests had been intermittently failing until yesterday, they should be more reliable in master now.

**Comment by travisn** [comment]
It seems the upgrade and smoke suites are failing consistently during test cleanup. The last thing in the test is:
```
2021-08-12 08:04:22.099486 I | testutil: Running kubectl [delete -f -]
objectbucketclaim.objectbucket.io "smoke-delete-bucket" deleted
panic: test timed out after 30m0s
```

The logs weren't collected unfortunately. It may be easiest to force push again and then connect to the tmate session to troubleshoot. The operator logs likely show some issue with the deletion. If it's your first time with the tmate session please do ping me or @subhamkrai.

**Comment by travisn** [comment]
You can ignore the warning from `kubectl apply`, the command still succeeded. 
Not sure why the latest lib-bucket-provisioner request would not be reaching the operator until the restart. @BlaineEXE Any ideas?

**Comment by BlaineEXE** [comment]
> You can ignore the warning from `kubectl apply`, the command still succeeded.
> Not sure why the latest lib-bucket-provisioner request would not be reaching the operator until the restart. @BlaineEXE Any ideas?

From the code, the delete operation is ignored because of the finalizer. I think this means the delete operations are handled via upates. With the new `updateSupported()` method coming before `ctrl.enqueueOBC()`, I suspect the OBC is not getting enqueued when the deletion timestamp changes.

A question I have is what **other** things might be getting skipped because of this? I think the `updateSupported()` method only applies when there are `spec` changes. I think if other things besides `spec` change, the update should be supported still.

Specifically, I think this (from the recent commit in lib-bucket-provisioner) is not correct:
```go
if reflect.DeepEqual(new.Spec, old.Spec) {
	return false
}
```
If the new spec is the same as the old spec, then something else has changed, like metadata, and the update should still be supported. We really only want to stop the update from happening if the spec is modified AND unsupported parts of the spec are modified.

When you were developing the lib-bucket-provisioner changes, were you testing them in Rook? You really should be doing that if not so we don't get broken commits to lib-bucket-provisioner.

**Comment by thotz** [comment]
> > You can ignore the warning from `kubectl apply`, the command still succeeded.
> > Not sure why the latest lib-bucket-provisioner request would not be reaching the operator until the restart. @BlaineEXE Any ideas?
> 
> From the code, the delete operation is ignored because of the finalizer. I think this means the delete operations are handled via upates. With the new `updateSupported()` method coming before `ctrl.enqueueOBC()`, I suspect the OBC is not getting enqueued when the deletion timestamp changes.
> 
> A question I have is what **other** things might be getting skipped because of this? I think the `updateSupported()` method only applies when there are `spec` changes. I think if other things besides `spec` change, the update should be supported still.
> 
> Specifically, I think this (from the recent commit in lib-bucket-provisioner) is not correct:
> 
> ```go
> if reflect.DeepEqual(new.Spec, old.Spec) {
> 	return false
> }
> ```
> 
> If the new spec is the same as the old spec, then something else has changed, like metadata, and the update should still be supported. We really only want to stop the update from happening if the spec is modified AND unsupported parts of the spec are modified.
> 
> When you were developing the lib-bucket-provisioner changes, were you testing them in Rook? You really should be doing that if not so we don't get broken commits to lib-bucket-provisioner.

@BlaineEXE : Thanks for debugging this issue, I added check to avoid update call if deletion stamp is set in NewController() via https://github.com/kube-object-storage/lib-bucket-provisioner/pull/216, it works fine. Let me know your thoughts on the same

**Comment by BlaineEXE** [comment]
Please update this PR to reference your `thotz:fixobcdeletestuck` branch so that the Rook CI will validate it also.

**Comment by thotz** [comment]
> Please update this PR to reference your `thotz:fixobcdeletestuck` branch so that the Rook CI will validate it also.

Sure

**Comment by BlaineEXE** [comment]
@thotz thanks. Merged the lib-bucket-provisioner fix. You should be good to go here.

**Comment by thotz** [comment]
> @thotz thanks. Merged the lib-bucket-provisioner fix. You should be good to go here.

Done

File: tests/integration/ceph_base_object_test.go [hunk 2]
```diff
@@ -279,12 +281,26 @@ func testObjectStoreOperations(s suite.Suite, helper *clients.TestClient, k8sh *
...
+		t.Run("test update quota on OBC bucket", func(t *testing.T) {
+			poErr := helper.BucketClient.UpdateObc(obcName, bucketStorageClassName, bucketname, newMaxObject, true)
+			assert.Nil(s.T(), poErr)
+			updated := utils.Retry(5, 2*time.Second, "OBC is created", func() bool {
+				return helper.BucketClient.CheckOBMaxObject(obcName, newMaxObject)
+			})
+			assert.True(s.T(), updated)
+			logger.Infof("Testing the updated object limit")
+			_, poErr = s3client.PutObjectInBucket(bucketname, ObjBody, ObjectKey3, contentType)
+			assert.Nil(s.T(), poErr)
```
Comment (BlaineEXE) [inline tests/integration/ceph_base_object_test.go:293]:
```suggestion
			assert.NoError(s.T(), poErr)
```

File: tests/integration/ceph_base_object_test.go [hunk 3]
```diff
@@ -279,12 +281,26 @@ func testObjectStoreOperations(s suite.Suite, helper *clients.TestClient, k8sh *
 			_, poErr = s3client.PutObjectInBucket(bucketname, ObjBody, ObjectKey3, contentType)
 			assert.Error(s.T(), poErr)
 		})
-
+		t.Run("test update quota on OBC bucket", func(t *testing.T) {
```
Comment (BlaineEXE) [inline tests/integration/ceph_base_object_test.go:284]: Insert newline before new test.

File: tests/integration/ceph_base_object_test.go [hunk 4]
```diff
@@ -279,12 +281,26 @@ func testObjectStoreOperations(s suite.Suite, helper *clients.TestClient, k8sh *
...
+				return helper.BucketClient.CheckOBMaxObject(obcName, newMaxObject)
+			})
+			assert.True(s.T(), updated)
+			logger.Infof("Testing the updated object limit")
+			_, poErr = s3client.PutObjectInBucket(bucketname, ObjBody, ObjectKey3, contentType)
+			assert.Nil(s.T(), poErr)
+			_, poErr = s3client.PutObjectInBucket(bucketname, ObjBody, ObjectKey4, contentType)
+			assert.Error(s.T(), poErr)
+		})
 		t.Run("delete objects on OBC bucket", func(t *testing.T) {
```
Comment (BlaineEXE) [inline tests/integration/ceph_base_object_test.go:297]: Newline before each test.

File: tests/integration/ceph_base_object_test.go [hunk 5]
```diff
@@ -279,12 +281,26 @@ func testObjectStoreOperations(s suite.Suite, helper *clients.TestClient, k8sh *
 			_, poErr = s3client.PutObjectInBucket(bucketname, ObjBody, ObjectKey3, contentType)
 			assert.Error(s.T(), poErr)
 		})
-
+		t.Run("test update quota on OBC bucket", func(t *testing.T) {
+			poErr := helper.BucketClient.UpdateObc(obcName, bucketStorageClassName, bucketname, newMaxObject, true)
+			assert.Nil(s.T(), poErr)
+			updated := utils.Retry(5, 2*time.Second, "OBC is created", func() bool {
```
Comment (BlaineEXE) [inline tests/integration/ceph_base_object_test.go:287]: Should this message be changed?

```suggestion
			updated := utils.Retry(5, 2*time.Second, "OBC is updated", func() bool {
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Just nit comments. This looks good overall. Thanks @thotz

**Comment by thotz** [comment]
> Just nit comments. This looks good overall. Thanks @thotz

@BlaineEXE : Done

**Review by BlaineEXE — APPROVED** [review]

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 6]
```diff
@@ -665,3 +665,69 @@ func (p *Provisioner) setAdminOpsAPIClient() error {
 
 	return nil
 }
+func (p Provisioner) updateAdditionalSettings(ob *bktv1alpha1.ObjectBucket) error {
+	var maxObjectsInt64 int64
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/provisioner.go:669]:
```suggestion
	var maxObjectsInt64, maxSizeInt64 int64
```

**Review by leseb — APPROVED** [review]
nit feel free to address or not

**Review by travisn — APPROVED** [review]

**Pull request merged**